proposals/compile-time-dependency-resolution.md

```diff
@@ -0,0 +1,489 @@
+
+#### Class constraint
+
+```kotlin
+extension class GroupRepository<A>(with R: Repository<A>) : Repository<Group<A>> {
+  /* ... */
+}
+```
+
+#### Function constraint
+
+```kotlin
+fun <A> fetch(id: String, with R: Repository<A>): A = loadById(id) // function position using parameter "R"
+```
+
+#### Extension evidence using an Object
+
+```kotlin
+extension object UserRepository: Repository<User> {
+  override fun loadAll(): List<User> {
+    return listOf(User(25, "Bob"))
+  }
+
+  override fun loadById(id: Int): User? {
+    return if (id == 25) {
+      User(25, "Bob")
+    } else {
+      null
+    }
+  }
+}
+```
+
+#### Extension evidence using a Class
+
+```kotlin
+extension class UserRepository: Repository<User> {
+  override fun loadAll(): List<User> {
+    return listOf(User(25, "Bob"))
+  }
+
+  override fun loadById(id: Int): User? {
+    return if (id == 25) {
+      User(25, "Bob")
+    } else {
+      null
+    }
+  }
+}
+```
+
+## Extension resolution order
+
+Classical interfaces only permit their implementation at the site of a type definition. Compiler extension resolution pattern typically relax this rule and **allow extension evidences be declared outside of the type definition**. When relaxing this rule it is important to preserve the coherency we take for granted with classical interfaces.
+
+For those reasons constraint interfaces must be declared in one of the following places (in strict resolution order):
+
+1. Arguments of the caller function.
+2. Companion object for the target type (User).
+3. Companion object for the constraint interface we're looking for (Repository).
+4. Subpackages of the package where the target type (User) to resolve is defined.
+5. Subpackages of the package where the constraint interface (Repository) is defined.
+
+All other instances are considered orphan instances and are not allowed. See [Appendix A](#Appendix-A) for a modification to this proposal that allows for orphan instances.
+
+Additionally, a constraint extension must not conflict with any other pre-existing extension for the same constraint interface; for the purposes of checking this we use the normal resolution rules. That's what we refer as compiler "coherence".
+
+#### 1. Arguments of the caller function
+
+It looks into the caller function argument list for an evidence of the required extension. Here, `bothValid()` gets a `Validator<A>` passed in so whenever it needs to resolve it for the inner calls to `validate()`, it'll be able to retrieve it from its own argument list.
+```kotlin
+fun <A> validate(a: A, with validator: Validator<A>): Boolean = a.isValid()
+
+fun <A> bothValid(x: A, y: A, with validator: Validator<A>): Boolean = validate(x) && validate(y)
+```
+
+#### 2. Companion object for the target type
+
+In case there's no evidence at the caller function level, we'll look into the companion of the target type. Let's say we have `Repository<A>` and `User` for the `A` type:
```

Comment:
```suggestion
In case there's no evidence at the caller function level, we'll look into the companion of the target type. Let's say we have an extension of `Validator<User>`:
```